x/tokenregistry/keeper/keeper.go

```diff
@@ -84,63 +84,70 @@ func (k keeper) GetEntry(wl types.Registry, denom string) (*types.RegistryEntry,
 	return nil, errors.Wrap(errors.ErrKeyNotFound, "registry entry not found")
 }
 
-func (k keeper) SetToken(ctx sdk.Context, entry *types.RegistryEntry) {
-	wl := k.GetRegistry(ctx)
+func (k keeper) GetDenomPrefix(ctx sdk.Context, denom string) []byte {
+	return append(types.TokenDenomPrefix, []byte(denom)...)
+}
 
-	for i := range wl.Entries {
-		if wl.Entries[i] != nil && strings.EqualFold(wl.Entries[i].Denom, entry.Denom) {
-			wl.Entries[i] = entry
-			k.SetRegistry(ctx, wl)
-			return
-		}
-	}
-	wl.Entries = append(wl.Entries, entry)
-	k.SetRegistry(ctx, wl)
+func (k keeper) GetDenom(ctx sdk.Context, denom string) types.RegistryEntry {
+
+	var entry types.RegistryEntry
+	key := k.GetDenomPrefix(ctx, denom)
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(key)
+
+	k.cdc.MustUnmarshal(bz, &entry)
+
+	return entry
 }
 
-func (k keeper) RemoveToken(ctx sdk.Context, denom string) {
-	registry := k.GetRegistry(ctx)
-	updated := make([]*types.RegistryEntry, 0)
-	for _, t := range registry.Entries {
-		if t != nil && !strings.EqualFold(t.Denom, denom) {
-			updated = append(updated, t)
-		}
-	}
-	k.SetRegistry(ctx, types.Registry{
-		Entries: updated,
-	})
+// SetToken add a new denom
+func (k keeper) SetToken(ctx sdk.Context, entry *types.RegistryEntry) {
+	entry.Sanitize()
+	key := k.GetDenomPrefix(ctx, entry.Denom)
+	store := ctx.KVStore(k.storeKey)
+
+	bz := k.cdc.MustMarshal(entry)
+
+	store.Set(key, bz)
 }
 
-func (k keeper) SetRegistry(ctx sdk.Context, wl types.Registry) {
+// RemoveToken remove a token
+func (k keeper) RemoveToken(ctx sdk.Context, denom string) {
+	key := k.GetDenomPrefix(ctx, denom)
 	store := ctx.KVStore(k.storeKey)
-	bz := k.cdc.MustMarshal(&wl)
-	store.Set(types.WhitelistStorePrefix, bz)
+	store.Delete(key)
 }
 
+// GetRegistry get all token's metadata
 func (k keeper) GetRegistry(ctx sdk.Context) types.Registry {
-	var whitelist types.Registry
+	var entries []*types.RegistryEntry
```

Comment:
This won't work for Peggy 2.0. It loops over every token correct? If so you have to assume there are millions of tokens imported so it must support pagination.

Reply:
so we use this pr to fix new ticket, or create new one?

Reply:
I created a new ticket and am going to take a look at it.

Reply:
Ticket is #2550 we can merge once everything else is done, will make this a separate PR.